src/common/contexts/provider.tsx

```diff
@@ -83,105 +81,78 @@ export const ProviderContextProvider: FunctionComponent<ProviderContextProviderP
-        console.warn(e.message);
-        await updateProviderOrSigner(createProvider(chainInfo.chainId));
-      } else {
-        throw e;
-      }
-    }
+    await walletSwitchChain(chainId);
+    setCurrentChainId(chainId);
   };
 
-  const getProvider = useCallback(() => {
-    if (providers.Provider.isProvider(providerOrSigner)) return providerOrSigner;
-    if (Signer.isSigner(providerOrSigner)) return providerOrSigner.provider;
-    return undefined;
-  }, [providerOrSigner]);
-
-  const getSigner = useCallback(
-    () => (isConnected ? (providerOrSigner as Signer) : undefined),
-    [isConnected, providerOrSigner]
-  );
-
-  const getTransactor = useCallback(() => getSigner() ?? getProvider(), [getProvider, getSigner]);
-
-  const getWeb3Provider = () => {
+  const updateProvider = useCallback(async () => {
     const { ethereum, web3 } = window;
     const metamaskExtensionNotFound = typeof ethereum === "undefined" || typeof web3 === "undefined";
-    if (metamaskExtensionNotFound || !ethereum.request) throw new NoMetaMaskError();
-
-    const injectedWeb3 = ethereum || web3.currentProvider;
-    if (!injectedWeb3) throw new Error("No injected web3 provider found");
-    return new ethers.providers.Web3Provider(injectedWeb3, "any");
-  };
+    if (metamaskExtensionNotFound || !ethereum.request) {
+      setProvider(createProvider(currentChainId || defaultChainId));
+      setAccount(undefined);
+    } else {
+      const injectedWeb3 = ethereum || web3.currentProvider;
+      if (!injectedWeb3) throw new Error("No injected web3 provider found");
+      const newProvider = new ethers.providers.Web3Provider(injectedWeb3, "any");
+      const network = await newProvider.getNetwork();
+      if (!isSupportedNetwork(network.chainId)) {
+        console.warn("User wallet is connected to an unsupported network, will fallback to default network");
+        setProvider(undefined);
+        setAccount(undefined);
+        setCurrentChainId(undefined);
+      } else {
+        setProvider(newProvider);
+        setCurrentChainId(network.chainId);
+      }
+    }
+  }, [currentChainId, defaultChainId, isSupportedNetwork]);
 
-  const requestSwitchChain = async (chainId: ChainId) => {
-    const { ethereum } = window;
-    if (!ethereum || !ethereum.request) return;
+  const updateSigner = useCallback(async () => {
     try {
-      await ethereum.request({
-        method: "wallet_switchEthereumChain",
-        params: [{ chainId: `0x${chainId.toString(16)}` }],
-      });
-    } catch (e: any) {
-      if (e.code === -32601) {
-        // Possibly on localhost which doesn't support the call
-        return console.error(e);
-      }
-      throw e;
+      const signer = (provider as ethers.providers.Web3Provider).getSigner();
+      const address = await signer.getAddress();
+      setAccount(address);
+      setProviderOrSigner(signer);
+    } catch (e) {
+      setAccount(undefined);
+      setProviderOrSigner(provider);
```

Comment:
thoughts: `updateSigner` yet `setProviderOrSigner`

Reply:
The `updateSigner` is in charge of updating the states for the `providerOrSigner` and `account`.